pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+// n is greater than 0.
+func newSeries(lset labels.Labels, n int) *series {
+	return &series{lset: lset, n: n}
+}
+
+// newSeries returns a new series and automatically cuts a new chunk after n samples if
+// n is greater than 0.
+// It extends the metric name of the label set by the aggregation modifier.
+func newAggrSeries(lset labels.Labels, aggr string, n int) *series {
+	return &series{lset: aggrLset(lset, aggr), n: n}
+}
+
+func (s *series) add(t int64, v float64) {
+	if s.n > 0 && s.l >= s.n {
+		s.cut()
+	}
+	if s.cur == nil {
+		s.cur = chunkenc.NewXORChunk()
+		s.cmin = t
+		s.app, _ = s.cur.Appender()
+	}
+	s.app.Append(t, v)
+	s.cmax = t
+}
+
+func (s *series) cut() {
+	s.chunks = append(s.chunks, chunks.Meta{
+		MinTime: s.cmin,
+		MaxTime: s.cmax,
+		Chunk:   s.cur,
+	})
+	s.cur = nil
+}
+
+func (s *series) close() {
+	if s.cur == nil {
+		return
+	}
+	s.chunks = append(s.chunks, chunks.Meta{
+		MinTime: s.cmin,
+		MaxTime: s.cmax,
+		Chunk:   s.cur,
+	})
+}
+
+// downsampleCounter downsamples the data points into the series with the given window.
+// If split is set to false, it will write all result data into a single chunk in the series.
+// It writes a signal sample to the end of each chunk which preservers the last true sample of
+// the input data.
+// If it is called on already downsampled data, the input data must end at such a signaling
+// sample to recursively guarantee the semantics.
+func downsampleCounter(ser *series, data []sample, window int64, split bool) {
+	var (
+		n       = targetChunkSize(len(data))
+		nextT   = int64(-1)
+		lastT   int64
+		lastV   float64
+		counter float64
+	)
+	for i, s := range data {
+		if s.t > nextT {
+			if split && i > 0 && i%n == 0 {
+				// Add signaling sample that indicates what the true last value was
+				// before cutting a new chunk.
+				// For already downsampled series lastV is set to the last value that
+				// series itself added as a signaling sample.
+				ser.add(nextT, lastV)
+				ser.cut()
+			}
+			if nextT != -1 {
+				ser.add(nextT, counter)
+			}
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		if i == 0 {
+			// First sample sets the counter.
+			counter = s.v
+		} else if s.t > lastT {
+			// We ignored samples with no increasing timestamps. This handles signaling samples
+			// in already downsampled counter series.
+			// At breaking points (e.g. chunks) they add a sample with non-increased timestamp
+			// and the true last value, like the one we add ourselves in this function.
+			if s.v < lastV {
+				// Counter reset, correct the value.
+				counter += s.v
+			} else {
+				// Add delta with last value to the counter.
+				counter += s.v - lastV
+			}
+		}
+		lastV = s.v
+		lastT = s.t
+	}
+	ser.add(nextT, counter)
+	// Add signaling sample that indicates what the true last value was.
+	ser.add(nextT, lastV)
+	ser.cut()
+}
+
+func downsampleCounterCounter(ser *series, data []sample, chunked [][]sample, window int64) {
```

Comment:
Maybe just `downsampleAggrCounter` . Also worth to add some comments to all of these functions (specifically what Counter suffix means and why it is useful to have it in aggr chunk)

Reply:
Also, do we really need to duplicate samples here? We have same samples in `data` and `chunked` can we optimize heap alloc by having `chunked` only or something? I am guessing that this will look too ugly

Reply:
Yes, better name!

`chunked` is just a set of sub-slices over `data`, so that should not be a significant amount of extra data. Also both slices are being reused in `Downsample`.

Reply:
ok, fine then - so just the name